### ISSUE TITLE
Allow for default values in configuration substitution

### ISSUE THREAD

**Issue by Pike** [backstage/backstage#14349]
When writing config yamls, one can do `${FOO}` to fill in values from the environment. It'd be nice to take that bash-ism a step further and allow for `${FOO:-default value}`.

## Feature Suggestion

Add a default/fallback logic to [config substitution](https://github.com/backstage/backstage/blob/master/packages/config-loader/src/lib/transform/substitution.ts), so that one could, for example, do

```yaml
app:
  title: My Backstage
  baseUrl: ${BASE_URL:-http://localhost:3000)
```

That'd be nice to substitute in CI for production URLs, and still have a dev config for localhost. While this can be worked around with `app-config.local.yaml`, those are usually really local, so don't allow for shared setups.

And they're also picked up in docker builds, IIRC.

## Context

For some reason, my deployed instance doesn't serve the production config. It's read and inserted, I see that. It's just not served.

**Comment by freben** [comment]
By served, do you mean in the frontend? If so, then bear in mind that you have to supply the same `--config` arguments to the production build command as during runtime! The reason is, the production config gets injected into the static frontend files during build - it's not served from the backend to the frontend. That quirk is something that we've been wanting to address, but haven't had the time to do so yet.

**Comment by freben** [comment]
That being said, I personally think that the feature request isn't bad as such, it could be neat in some circumstances. Maybe it does introduce risk of over-using that pattern instead of proper config file layering, though; that's what you normally should lean against.

**Comment by Pike** [comment]
I am seeing the `Injecting env config into module-backstage.8f130c7a.js` from https://github.com/backstage/backstage/blob/b725ae2bc70cf9be3e9630df484daae1fb88e8f5/plugins/app-backend/src/lib/config.ts#L48 in the startup logs in the deployment, after `Loaded config from app-config.yaml, app-config.production.yaml`. So I think this should work, but somehow it doesn't. It does on the same docker image locally, but not in deployment. Or its intermittent, 'cause what I'm currently doing doesn't do anything either. Sorry, this sounds really confusing, and it is to me.

As for `--config`, I was looking into that and turns out that the top-level `yarn build` aka `backstage-cli repo build --all` doesn't like the `--config` options, so I dropped that in favor of mocking a `app-config.local.yaml` with the relevant `app-config.production.yaml` parts as part of CI.

**Comment by Rugvip** [comment]
Regarding the lack of `--config` for the top-level build, that was an oversight and is being fixed #14329 😅
